### ISSUE TITLE
Random seeds are used for training

### ISSUE THREAD

**Issue by sunshehai** [CityU-AIM-Group/SIGMA#41]
@wymanCV 
In your code, random seeds are used for training, so code based on Faster RCNN cannot be reproduced, the result will fluctuate, and code based on FCOS can be reproduced?  My dataset is like this, thank you.

**Comment by wymanCV** [comment]
I have provided the training log for the faster rcnn implementation, which can reproduce the results. The random seed doesn't make much influence for the results since I didn't adjust it.

**Comment by sunshehai** [comment]
@wymanCV 
In your code FRCNN-SIGMA ++, random seed does not work. In other domains based on Faster RCNN to adapt target detection, random seed also does not work. I would like to ask how to make random seed function in the code FRCNN-SIGMA ++? Thank you.
In my opinion, the optimal model can be selected through verification every 100 iterations, and the reproducibility of each result can be guaranteed if random seeds work. However, it is a pity that the current domain adaptive target detection based on Faster RCNN cannot fix random seeds.
In addition, in your code FRCNN-SIGMA ++, I found that the position of random seeds was changed from that found before, but I changed back to the original position, the random seeds did not work either.
I mainly found that random seeds could not be fixed in the experiment, sometimes the results would fluctuate greatly, and sometimes the results would be the same. Thank you very much for your reply.

**Comment by sunshehai** [comment]
@wymanCV 
Thank you for your reply. I used a single GPU for training.
1.Is it best to set the number of iterations to 1 epoch according to my own dataset? 

2.Cityscapes have 2975 train images, your "IMS_PER_BATCH" is 2, so maybe 2000-iteration evaluation for the FRCNN implementation? 

3.In addition, I noticed that your code has different statements about "IMS_PER_BATCH". In fact, the number of source or target domains is about half of "IMS_PER_BATCH"? Thank you.

**Comment by wymanCV** [comment]
1. Setting training iterations as one epoch for the custom dataset is far not enough. The 4x training schedule may be better, i.e., around 4x12 epochs. Moreover, it depends on the scale of your dataset. For example, you may need more epochs for small-scale datasets. You may need to do more experiments for verification on custom datasets. This is different from fully-supervised learning since it may even cause a performance decrease if the training is too short and long on some datasets.
2. Evaluating per-epoch is ok. But as far as I know, performing validation more frequently can better ensure the reprobubalilty in most open-sourced DAOD works. Otherwise, there will be inevitable performance differences and the need to conduct an offline model selection. 
3.  Sorry for that small mistake on the code comment. IMS_PER_BATCH is for the number of per-domain images for FCOS, which differs from the FRCNN (both-domain images). The differences between the two frameworks cause this inconsistency.

Thank you very much for your interest in our work. Feel free to let me know if you have other issues.

**Comment by wymanCV** [comment]
Thanks for your carefulness!

**Comment by sunshehai** [comment]
@wymanCV  I encountered the following problem when porting the model selection procedure
1.In your SIGMA model, https://github.com/CityU-AIM-Group/SIGMA/blob/main/fcos_core/engine/inference.py，line 29，targets=targets. However, in EPM model, targets=None, why is different?
2.In your SIGMA model, https://github.com/CityU-AIM-Group/SIGMA/blob/main/fcos_core/engine/validation.py, line 55,
targets=None, are targets not label? Why can we verify each time without label? Thank you.

**Comment by wymanCV** [comment]
Hi, thanks for your carefulness!

That doesn't make difference since the module focs_head doesn't use targets. I sent targets to focus_head just for the debugging purpose in this project. You can turn it None as EPM and it won't make any influence.

The targets don't be used for evaluation. The true evaluation process is here

https://github.com/CityU-AIM-Group/SIGMA/blob/3775d869914adc74e52858b8cb89b9de875d5295/fcos_core/data/datasets/evaluation/__init__.py#L29

**Comment by sunshehai** [comment]
@wymanCV  Thank you for your reply.
https://github.com/CityU-AIM-Group/SIGMA/blob/FRCNN-SIGMA%2B%2B/fcos_core/data/build.py, line 17, is_source=False.
1.When domain adaptation is enabled,  target domain validation set; 2.When the target domain is closed, training set; 3.The data set when the training is over and the test is performed. All of them are (is_source=False). None of all are consistent with Model DA-FRCNN.
https://github.com/krumo/Domain-Adaptive-Faster-RCNN-PyTorch/blob/master/maskrcnn_benchmark/data/build.py
https://github.com/krumo/Domain-Adaptive-Faster-RCNN-PyTorch/blob/master/tools/train_net.py

**Comment by sunshehai** [comment]
1.val_data_loader, 2.data_loader, 3.data_loaders_val, they are all (is_source=False).  I think the result may be (is_source=True).

**Comment by wymanCV** [comment]
You'd better read the code more carefully :)

source: see here
https://github.com/CityU-AIM-Group/SIGMA/blob/23bfcc3b6ca6c6320fa5b9c5295b99c4a46e4374/fcos_core/data/build.py#L229

target see here 
https://github.com/CityU-AIM-Group/SIGMA/blob/23bfcc3b6ca6c6320fa5b9c5295b99c4a46e4374/fcos_core/data/build.py#L252

is_source=False by default, see here
https://github.com/CityU-AIM-Group/SIGMA/blob/23bfcc3b6ca6c6320fa5b9c5295b99c4a46e4374/fcos_core/data/build.py#L17

**Comment by wymanCV** [comment]
> 1.val_data_loader, 2.data_loader, 3.data_loaders_val, they are all (is_source=False). I think the result may be (is_source=True).

That doesn't have influence in UDA training and inference. Thanks.

**Comment by sunshehai** [comment]
Ok, thank you very much for your reply. I didn't understand something in the verification set just now. Thank you very much for your patience.

**Comment by wymanCV** [comment]
> Ok, thank you very much for your reply. I didn't understand something in the verification set just now. Thank you very much for your patience.

You can print the image information for each dataloader to check. Some details of the project didn't strictly follow other projects, e.g., DAF and EPM, but has the same function as them in UDA.

**Comment by sunshehai** [comment]
Thank you very much for helping me learn a lot through your code and patience. Thank you.
